src/Resources/ResourceManager/Implementation/Deployments/RemoveAzureDeploymentCmdlet.cs

```diff
@@ -69,17 +86,91 @@ public override void ExecuteCmdlet()
+                    var deploymentName = this.GetDeploymetName();
+
+                    var managementGroupId = this.GetManagementGroupId();
+                    var resourceGroupName = this.GetResourceGroupName();
 
-                    ResourceManagerSdkClient.DeleteDeploymentAtSubscriptionScope(deploymentName);
+                    this.DeleteDeployment(managementGroupId, resourceGroupName, deploymentName);
 
                     if (this.PassThru.IsPresent)
                     {
                         WriteObject(true);
                     }
                 });
         }
+
+        private string GetDeploymetName()
+        {
+            if (!string.IsNullOrEmpty(this.Name))
+            {
+                return this.Name;
+            }
+            else if (!string.IsNullOrEmpty(this.Id))
+            {
+                return ResourceIdUtility.GetResourceName(this.Id);
+            }
+            else
+            {
+                return this.InputObject.DeploymentName;
+            }
+        }
+
+        private string GetManagementGroupId()
+        {
+            if (!string.IsNullOrEmpty(this.ManagementGroupId))
+            {
+                return this.ManagementGroupId;
+            }
+            else if (!string.IsNullOrEmpty(this.Id))
+            {
+                return ResourceIdUtility.GetManagementGroupId(this.Id);
+            }
+            else if (this.InputObject != null)
+            {
+                return this.InputObject.ManagementGroupId;
+            }
+
+            return null;
+        }
+
+        private string GetResourceGroupName()
+        {
+            if (!string.IsNullOrEmpty(this.ResourceGroupName))
+            {
+                return this.ResourceGroupName;
+            }
+            else if (!string.IsNullOrEmpty(this.Id))
+            {
+                return ResourceIdUtility.GetResourceGroupName(this.Id);
+            }
+            else if (this.InputObject != null)
+            {
+                return this.InputObject.ResourceGroupName;
+            }
+
+            return null;
+        }
+
+        private void DeleteDeployment(string managementGroupId, string resourceGroupName, string deploymentName)
+        {
+            if (this.Tenant)
+            {
+                // (tiano): parse Id for tenant deployment too. 
+                ResourceManagerSdkClient.DeleteDeploymentAtTenantScope(deploymentName);
+            }
+            else if (!string.IsNullOrEmpty(managementGroupId))
+            {
+                ResourceManagerSdkClient.DeleteDeploymentAtManagementGroup(managementGroupId, deploymentName);
+            }
+            else if (!string.IsNullOrEmpty(resourceGroupName))
+            {
+                ResourceManagerSdkClient.DeleteDeploymentAtResourceGroup(resourceGroupName, deploymentName);
+            }
+            else
+            {
+                ResourceManagerSdkClient.DeleteDeploymentAtSubscriptionScope(deploymentName);
+            }
```

Comment:
Not necessarily an issue, but I'm seeing this switch pattern in a bunch of places. Would it make sense to capture it in a method that takes a tenant-level delegate, a mg-level delegate, a sub-level delegate, and an RG-level delegate? With named parameters, it might help readability. 

On the other hand, once you figure it out one of them, it's obvious what this does. What do you think?

Reply:
I'm usually not a big fan of passing delegates to functions unless it really simplifies implementation or improves readability.  I can take a look and see if that's the case here.

Reply:
Sounds good. I'm ok either way.

---
In reply to: [308965056](https://github.com/Azure/azure-powershell/pull/9707#discussion_r308965056) [](ancestors = 308965056)